optaplanner-core/src/main/java/org/optaplanner/core/impl/util/IntervalSplitPoint.java

```diff
@@ -0,0 +1,131 @@
+                        .thenComparingInt(interval -> System.identityHashCode(interval.value)));
+        intervalsEndingAtSplitPointSet = new TreeSet<>(
+                Comparator.<Interval<_IntervalValue, _PointValue>, _PointValue> comparing(Interval::getStart)
+                        .thenComparingInt(interval -> System.identityHashCode(interval.value)));
+    }
+
+    public boolean addIntervalStartingAtSplitPoint(Interval<_IntervalValue, _PointValue> interval) {
+        startIntervalToCountMap.merge(interval.value, 1, Integer::sum);
+        return intervalsStartingAtSplitPointSet.add(interval);
+    }
+
+    public void removeIntervalStartingAtSplitPoint(Interval<_IntervalValue, _PointValue> interval) {
+        Integer newCount = startIntervalToCountMap.computeIfPresent(interval.value, (key, count) -> {
+            if (count > 1) {
+                return count - 1;
+            }
+            return null;
+        });
+        if (null == newCount) {
+            intervalsStartingAtSplitPointSet.remove(interval);
+        }
+    }
+
+    public boolean addIntervalEndingAtSplitPoint(Interval<_IntervalValue, _PointValue> interval) {
+        endIntervalToCountMap.merge(interval.value, 1, Integer::sum);
+        return intervalsEndingAtSplitPointSet.add(interval);
+    }
+
+    public void removeIntervalEndingAtSplitPoint(Interval<_IntervalValue, _PointValue> interval) {
+        Integer newCount = endIntervalToCountMap.computeIfPresent(interval.value, (key, count) -> {
+            if (count > 1) {
+                return count - 1;
+            }
+            return null;
+        });
+        if (null == newCount) {
+            intervalsEndingAtSplitPointSet.remove(interval);
+        }
+    }
+
+    public boolean containsIntervalStarting(Interval<_IntervalValue, _PointValue> interval) {
+        return intervalsStartingAtSplitPointSet.contains(interval);
+    }
+
+    public boolean containsIntervalEnding(Interval<_IntervalValue, _PointValue> interval) {
+        return intervalsEndingAtSplitPointSet.contains(interval);
+    }
+
+    public Iterator<_IntervalValue> getValuesStartingFromSplitPointIterator() {
+        return intervalsStartingAtSplitPointSet.stream()
+                .flatMap(interval -> IntStream.range(0, startIntervalToCountMap.get(interval.value))
+                        .mapToObj((index) -> interval.value))
+                .iterator();
+    }
+
+    public boolean isEmpty() {
+        return intervalsStartingAtSplitPointSet.isEmpty() && intervalsEndingAtSplitPointSet.isEmpty();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o)
+            return true;
+        if (o == null || getClass() != o.getClass())
+            return false;
+        IntervalSplitPoint<?, ?> that = (IntervalSplitPoint<?, ?>) o;
+        return splitPoint.equals(that.splitPoint);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(splitPoint);
+    }
```

Comment:
Same question on the need for equality.

Reply:
`IntervalSplitPoint` implements `Comparable` so it should override equals IMO